rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -1939,33 +1939,33 @@ bool spu_thread::process_mfc_cmd()
 					}
 				}
 			}
-			else if (auto& data = vm::_ref<decltype(rdata)>(addr); rtime == (vm::reservation_acquire(raddr, 128) & -128) && cmp_rdata(rdata, data))
+			else if (auto& data = vm::_ref<decltype(rdata)>(addr); rtime == (vm::reservation_acquire(raddr, 128) & -128))
 			{
-				auto& res = vm::reservation_lock(raddr, 128);
-				const u64 old_time = res.load() & -128;
-
-				if (rtime == old_time)
+				if (cmp_rdata(rdata, to_write))
 				{
-					*reinterpret_cast<atomic_t<u32>*>(&data) += 0;
-
-					// Full lock (heavyweight)
-					// TODO: vm::check_addr
-					vm::writer_lock lock(addr);
-
-					if (cmp_rdata(rdata, data))
-					{
-						mov_rdata(data, to_write);
-						res.release(old_time + 128);
-						result = 1;
-					}
-					else
-					{
-						res.release(old_time);
-					}
+					// Pseudo write back: Check memory change without lock
+					result = cmp_rdata(rdata, data) && vm::reservation_acquire(raddr, 128).compare_and_swap_test(rtime, rtime + 128);
 				}
 				else
 				{
-					res.release(old_time);
+					auto& res = vm::reservation_lock(raddr, 128);
+					const u64 old_time = res.load() & -128;
+
+					if (rtime == old_time)
+					{
+						*reinterpret_cast<atomic_t<u32>*>(&data) += 0;
+
+						// Full lock (heavyweight)
+						// TODO: vm::check_addr
+						vm::writer_lock lock(addr);
+
+						if (cmp_rdata(rdata, data))
+						{
+							mov_rdata(data, to_write);
+							result = 1;
+						}
+					}
+					res.release(old_time + (long long)(result << 7));
```

Comment:
Don't use C style cast. It could be `u64{result << 7}`

Reply:
There's no need for a cast here.. just parenthesses are enough.